Add CSRF protection using hapi/crumb

jakubdatawrapper opened the pull request:
#### Motivation

Mitigate CSRF.

Issue: https://www.notion.so/datawrapper/Fix-CSRF-vulnerabilities-320648eba56144bdb914a1448b04f04c

Pentest issue no. 7.

#### Changes

1. Send a cookie and a matching `X-CSRF-Token` header with all state-changing (aka "unsafe", ie. POST, DELETE...) requests authenticated using a session cookie. Then check that the header and the cookie match. Use `hapi/crumb` for that.

    https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#double-submit-cookie

2. Check the Referer header for all state-changing (aka "unsafe", ie. POST, DELETE...) requests authenticated using a session cookie.

    https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#verifying-origin-with-standard-headers

    The implementation is modeled upon the one in the Django framework: https://github.com/django/django/blob/master/django/middleware/csrf.py#L224

    Add a unit test to the `me/index.test.js` test suite.

The following routes are exempt from the CSRF check, because they are called before a cookie can be set:

- `POST /v3/auth/login`
- `POST /v3/auth/session`

#### Remaining issues

- [X] Will currently logged-in users get logged out when this change is deployed? -> **no, we will make a small GET request if the cookie is not set**
- [ ] What are the security implications of the login and session routes being exempt from the CSRF check?
- [X] Check how this behaves when using login tokens and embeds on third-party URLs. -> **works**
- [x] Depends on https://github.com/datawrapper/api/pull/218
- [x] Depends on https://github.com/datawrapper/shared/pull/38
- [x] Depends on https://github.com/datawrapper/chart-core/pull/47
- [x] Depends on https://github.com/datawrapper/controls/pull/94
- [x] Depends on https://github.com/datawrapper/frontend/pull/29
- [x] Depends on https://github.com/datawrapper/datawrapper/pull/395

davidkokkelink commented on src/server.js:
```diff
@@ -144,17 +173,40 @@ async function getVersionInfo() {
+                cookieOptions: {
+                    domain: '.' + config.api.domain,
+                    isHttpOnly: false,
+                    isSecure: !DW_DEV_MODE
```
This should use `config.frontend.https` rather than DW_DEV_MODE

davidkokkelink replied:
```suggestion
                    isSecure: config.frontend.https
```

davidkokkelink commented on src/server.js:
```diff
@@ -144,17 +173,40 @@ async function getVersionInfo() {
+        {
+            plugin: Crumb,
+            options: {
+                cookieOptions: {
+                    domain: '.' + config.api.domain,
+                    isHttpOnly: false,
```
Maybe we need to set the `SameSite` attribute here, similar to how we do in the core [getStateOpts method](https://github.com/datawrapper/api/blob/master/src/auth/utils.js#L126-L139)?

We should definitely check how this behaves when using login tokens and embeds on third-party URLs.

jakubdatawrapper replied:
I've added the sameSite and added a TODO to the PR description for the login tokens and embeds check.

jakubdatawrapper commented:
1. I tested the login token workflow and it works well.

2. I also made sure that if a user is logged in and the CSRF cookie is not set and we turn the requirement for CSRF cookies on, then httpReq makes a small GET request to `/v3/me` to trigger the setting of the cookie. See https://github.com/datawrapper/shared/blob/d9bd2c095007828bff6abcdbed1e8157c49d9a43/httpReq.js#L77

So from my point of view, we can start deploying the PRs. I will un-draft the first few and assign you, @davidkokkelink, as the reviewer.

sto3psl commented:
Hey @jakubdatawrapper, can you remove me from the Reviewers please 😄. I probably not the right person anymore 😉.

jakubdatawrapper commented:
Haha, hi Fabian :)

jakubdatawrapper commented:
This is now deployed on staging and works. I'd like to merge it to master.

I did two changes here today:
- Removed the dummy CSRF cookie from service-utils: https://github.com/datawrapper/service-utils/commit/76e372932d0645ac759e69f417ca2b60b15cfcaf
- Added hapi/crumb's built-in logging of failed CSRF token validations: https://github.com/datawrapper/api/pull/215/commits/d1bdbd6d3e652f56fbda387bc5bbb61770048653
